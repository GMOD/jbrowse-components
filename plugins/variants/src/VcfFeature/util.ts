import { parseBreakend } from '@gmod/vcf'
import type VCF from '@gmod/vcf'

const altTypeToSO: Record<string, string> = {
  DEL: 'deletion',
  INS: 'insertion',
  DUP: 'duplication',
  INV: 'inversion',
  INVDUP: 'inverted_duplication',
  CNV: 'copy_number_variation',
  TRA: 'translocation',
  'DUP:TANDEM': 'tandem_duplication',
  NON_REF: 'sequence_variant',
  '*': 'sequence_variant',
}

/**
 * Get a sequence ontology (SO) term that describes the variant type
 */
export function getSOTermAndDescription(
  ref: string,
  alt: string[] | undefined,
  parser: VCF,
): string[] {
  // it's just a remark if there are no alternate alleles
  if (!alt || alt.length === 0) {
    return ['remark', 'no alternative alleles']
  }

  const soTerms = new Set<string>()
  let descriptions = new Set<string>()
  alt.forEach(a => {
    let [soTerm, description] = getSOAndDescFromAltDefs(a, parser)
    if (!soTerm) {
      ;[soTerm, description] = getSOAndDescByExamination(ref, a)
    }
    if (soTerm && description) {
      soTerms.add(soTerm)
      descriptions.add(description)
    }
  })

  // Combine descriptions like ["SNV G -> A", "SNV G -> T"] to ["SNV G -> A,T"]
  if (descriptions.size > 1) {
    const descs = [...descriptions]
    const prefixes = new Set(
      descs
        .map(desc => {
          const prefix = desc.split('->')
          return prefix[1] ? prefix[0] : desc
        })
        .filter((f): f is string => !!f),
    )

    descriptions = new Set(
      [...prefixes]
        .map(r => r.trim())
        .map(prefix => {
          const suffixes = descs
            .map(desc => desc.split('->').map(r => r.trim()))
            .map(pref => (pref[1] && pref[0] === prefix ? pref[1] : ''))
            .filter(f => !!f)

          return suffixes.length ? `${prefix} -> ${suffixes.join(',')}` : prefix
        }),
    )
  }
  if (soTerms.size) {
    return [[...soTerms].join(','), [...descriptions].join(',')]
  }
  return []
}

export function getSOAndDescFromAltDefs(alt: string, parser: VCF): string[] {
  if (typeof alt === 'string' && !alt.startsWith('<')) {
    return []
  }

  // look for a definition with an SO type for this
  let soTerm = altTypeToSO[alt]
  // if no SO term but ALT is in metadata, assume sequence_variant
  if (!soTerm && parser.getMetadata('ALT', alt)) {
    soTerm = 'sequence_variant'
  }
  if (soTerm) {
    return [soTerm, alt]
  }

  // try to look for a definition for a parent term if we can
  const modAlt = alt.split(':')
  if (modAlt.length > 1) {
    return getSOAndDescFromAltDefs(`<${modAlt.slice(0, -1).join(':')}>`, parser)
  }

  // no parent
  return []
}

// note: term SNV is used instead of SNP because SO definition of SNP says
// abundance must be at least 1% in population, and can't be sure we meet
// that
export function getSOAndDescByExamination(ref: string, alt: string) {
  const bnd = parseBreakend(alt)
  if (bnd) {
    return ['breakend', alt]
  }
  if (ref.length === 1 && alt.length === 1) {
    return ['SNV', makeDescriptionString('SNV', ref, alt)]
  }
  if (alt === '<INS>') {
    return ['insertion', alt]
  }
  if (alt === '<DEL>') {
    return ['deletion', alt]
  }
  if (alt === '<DUP>') {
    return ['duplication', alt]
  }
  if (alt === '<CNV>') {
    return ['cnv', alt]
  }
  if (alt === '<INV>') {
    return ['inversion', alt]
  }
  if (alt === '<TRA>') {
    return ['translocation', alt]
  }
  if (alt.includes('<')) {
    return ['sv', alt]
  }
  if (ref.length === alt.length) {
    return ref.split('').reverse().join('') === alt
      ? ['inversion', makeDescriptionString('inversion', ref, alt)]
      : ['substitution', makeDescriptionString('substitution', ref, alt)]
  }
  if (ref.length <= alt.length) {
    const len = alt.length - ref.length
    const lena = len.toLocaleString('en-US')
    return [
      'insertion',
      len > 5 ? `${lena}bp INS` : makeDescriptionString('insertion', ref, alt),
    ]
  }
  if (ref.length > alt.length) {
    const len = ref.length - alt.length
    const lena = len.toLocaleString('en-US')
    return [
      'deletion',
      len > 5 ? `${lena}bp DEL` : makeDescriptionString('deletion', ref, alt),
    ]
  }

  return ['indel', makeDescriptionString('indel', ref, alt)]
}

function makeDescriptionString(soTerm: string, ref: string, alt: string) {
  return `${soTerm} ${ref} -> ${alt}`
}
