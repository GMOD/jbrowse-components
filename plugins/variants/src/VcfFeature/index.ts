import { getSOTermAndDescription } from './util'
import type VCF from '@gmod/vcf'
import type { Feature } from '@jbrowse/core/util'

// locals

type Samples = Record<
  string,
  Record<string, { values: string[] | number[] | null }>
>

interface FeatureData {
  [key: string]: unknown
  refName: string
  start: number
  end: number
  description?: string
  type?: string
  name?: string
  aliases?: string[]
  samples?: Samples
}

export default class VCFFeature implements Feature {
  private variant: any

  private parser: VCF

  private data: FeatureData

  private _id: string

  constructor(args: { variant: any; parser: VCF; id: string }) {
    this.variant = args.variant
    this.parser = args.parser
    this.data = this.dataFromVariant(this.variant)
    this._id = args.id
  }

  get(field: string): any {
    return field === 'samples'
      ? this.variant.SAMPLES
      : (this.data[field] ?? this.variant[field])
  }

  set() {}

  parent() {
    return undefined
  }

  children() {
    return undefined
  }

  tags() {
    return [...Object.keys(this.data), ...Object.keys(this.variant), 'samples']
  }

  id() {
    return this._id
  }

  dataFromVariant(variant: {
    REF: string
    POS: number
    ALT?: string[]
    CHROM: string
    INFO: any
    ID?: string[]
  }): FeatureData {
    const { REF, ALT, POS, CHROM, INFO, ID } = variant
    const start = POS - 1
    const [type, description] = getSOTermAndDescription(REF, ALT, this.parser)
    const isTRA = ALT?.includes('<TRA>')
    const isSymbolic = ALT?.some(f => f.includes('<'))

    return {
      refName: CHROM,
      start,
      end: isSymbolic && INFO.END && !isTRA ? +INFO.END[0] : start + REF.length,
      description,
      type,
      name: ID?.join(','),
      aliases: ID && ID.length > 1 ? ID.slice(1) : undefined,
    }
  }

  toJSON(): any {
    return {
      uniqueId: this._id,
      ...this.variant,
      ...this.data,
      samples: this.variant.SAMPLES,
    }
  }
}
