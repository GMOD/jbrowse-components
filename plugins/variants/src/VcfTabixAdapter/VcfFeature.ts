import { Feature } from '@jbrowse/core/util/simpleFeature'

/* eslint-disable no-underscore-dangle, @typescript-eslint/camelcase */

/**
 * VCF Feature creation with lazy genotpye evaluation.
 */
interface Samples {
  [key: string]: {
    [key: string]: { values: string[] | number[] | null }
  }
}
interface Breakend {
  MateDirection: string
  Replacement: string
  MatePosition: string
  Join: string
}
interface FeatureData {
  [key: string]: unknown
  refName: string
  start: number
  end: number
  description?: string
  type?: string
  name?: string
  aliases?: string[]
  samples?: Samples
}

export default class VCFFeature implements Feature {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private variant: any

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private parser: any

  private data: FeatureData

  private _id: string

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  constructor(args: { variant: any; parser: any; id: string }) {
    this.variant = args.variant
    this.parser = args.parser
    this.data = this.dataFromVariant(this.variant)
    this._id = args.id
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  get(field: string): any {
    if (field === 'samples') return this.variant.SAMPLES
    return this.variant[field] || this.data[field]
  }

  set(): void {}

  parent(): undefined {
    return undefined
  }

  children(): undefined {
    return undefined
  }

  tags(): string[] {
    const t = [
      ...Object.keys(this.data),
      ...Object.keys(this.variant),
      'samples',
    ]
    return t
  }

  id(): string {
    return this._id
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  dataFromVariant(variant: any): FeatureData {
    const start = variant.POS - 1
    const [SO_term, description] = this._getSOTermAndDescription(
      variant.REF,
      variant.ALT,
    )
    const isTRA =
      variant.ALT && variant.ALT.some((f: string | Breakend) => f === '<TRA>')
    const isSymbolic =
      variant.ALT &&
      variant.ALT.some(
        (f: string | Breakend) =>
          typeof f === 'string' && f.indexOf('<') !== -1,
      )
    const featureData: FeatureData = {
      refName: variant.CHROM,
      start,
      end:
        isSymbolic && variant.INFO.END && !isTRA
          ? Number(variant.INFO.END[0])
          : start + variant.REF.length,
      description,
      type: SO_term,
      name: variant.ID ? variant.ID[0] : undefined,
      aliases:
        variant.ID && variant.ID.length > 1 ? variant.ID.slice(1) : undefined,
    }

    return featureData
  }

  /**
   * Get a sequence ontology (SO) term that describes the variant type
   */
  _getSOTermAndDescription(
    ref: string,
    alt: string[],
  ): [string, string] | [undefined, undefined] {
    // it's just a remark if there are no alternate alleles
    if (!alt || alt === []) {
      return ['remark', 'no alternative alleles']
    }

    const soTerms: Set<string> = new Set()
    let descriptions: Set<string> = new Set()
    alt.forEach(a => {
      let [soTerm, description] = this._getSOAndDescFromAltDefs(ref, a)
      if (!soTerm) {
        ;[soTerm, description] = this._getSOAndDescByExamination(ref, a)
      }
      if (soTerm && description) {
        soTerms.add(soTerm)
        descriptions.add(description)
      }
    })
    // Combine descriptions like ["SNV G -> A", "SNV G -> T"] to ["SNV G -> A,T"]
    if (descriptions.size > 1) {
      const prefixes: Set<string> = new Set()
      descriptions.forEach(desc => {
        const prefix = /(\w+? \w+? -> )(?:<)\w+(?:>)/.exec(desc)
        if (prefix && prefix[1]) prefixes.add(prefix[1])
        else prefixes.add(desc)
      })
      const new_descs: string[] = []
      ;[...prefixes].forEach((prefix: string) => {
        const suffixes: string[] = []
        ;[...descriptions].forEach((desc: string) => {
          if (desc.startsWith(prefix)) {
            suffixes.push(desc.slice(prefix.length))
          }
        })
        new_descs.push(prefix + suffixes.join(','))
      })
      descriptions = new Set(new_descs)
    }
    if (soTerms.size) {
      return [[...soTerms].join(','), [...descriptions].join(',')]
    }
    return [undefined, undefined]
  }

  static _altTypeToSO: { [key: string]: string | undefined } = {
    DEL: 'deletion',
    INS: 'insertion',
    DUP: 'duplication',
    INV: 'inversion',
    INVDUP: 'inverted duplication',
    CNV: 'copy_number_variation',
    TRA: 'translocation',
    'DUP:TANDEM': 'tandem_duplication',
    NON_REF: 'sequence_variant',
    '*': 'sequence_variant',
  }

  _getSOAndDescFromAltDefs(
    ref: string,
    alt: string | Breakend,
  ): [string, string] | [undefined, undefined] {
    // not a symbolic ALT if doesn't begin with '<', so we'll have no definition
    if (typeof alt === 'object') {
      return [
        'breakend',
        'A VCF breakend defines one end of a point of structural variation',
      ]
    }

    if (typeof alt === 'string' && !alt.startsWith('<')) {
      return [undefined, undefined]
    }

    alt = alt.replace(/^<|>$/g, '') // trim off < and >

    // look for a definition with an SO type for this
    let soTerm = VCFFeature._altTypeToSO[alt]
    // if no SO term but ALT is in metadata, assume sequence_variant
    if (!soTerm && this.parser.getMetadata('ALT', alt))
      soTerm = 'sequence_variant'
    if (soTerm) {
      const description = this.parser.getMetadata('ALT', alt, 'Description')
        ? `${alt} - ${this.parser.getMetadata('ALT', alt, 'Description')}`
        : this._makeDescriptionString(soTerm, ref, alt)
      return [soTerm, description]
    }

    // try to look for a definition for a parent term if we can
    const modAlt = alt.split(':')
    if (modAlt.length > 1) {
      return this._getSOAndDescFromAltDefs(
        ref,
        `<${modAlt.slice(0, modAlt.length - 1).join(':')}>`,
      )
    }
    // no parent
    return [undefined, undefined]
  }

  _getSOAndDescByExamination(
    ref: string,
    alt: string | Breakend,
  ): [string, string] {
    if (typeof alt === 'object') {
      return ['breakend', this._makeDescriptionString('breakend', ref, alt)]
    }
    if (ref.length === 1 && alt.length === 1) {
      // use SNV because SO definition of SNP says abundance must be at
      // least 1% in population, and can't be sure we meet that
      return ['SNV', this._makeDescriptionString('SNV', ref, alt)]
    }
    if (ref.length === alt.length)
      if (ref.split('').reverse().join('') === alt)
        return ['inversion', this._makeDescriptionString('inversion', ref, alt)]
      else
        return [
          'substitution',
          this._makeDescriptionString('substitution', ref, alt),
        ]

    if (ref.length <= alt.length)
      return ['insertion', this._makeDescriptionString('insertion', ref, alt)]

    if (ref.length > alt.length)
      return ['deletion', this._makeDescriptionString('deletion', ref, alt)]

    return ['indel', this._makeDescriptionString('indel', ref, alt)]
  }

  _makeDescriptionString(
    soTerm: string,
    ref: string,
    alt: string | Breakend,
  ): string {
    return `${soTerm} ${ref} -> ${alt}`
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  toJSON(): any {
    return {
      uniqueId: this._id,
      ...this.data,
      ...this.variant,
      samples: this.variant.SAMPLES,
    }
  }
}
