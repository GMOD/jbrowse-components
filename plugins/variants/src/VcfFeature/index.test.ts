import VcfParser from '@gmod/vcf'
import VcfFeature from './index'

test('test usage of the VcfFeature', () => {
  const parser = new VcfParser({
    header:
      '#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tBAMs/caudaus.sorted.sam',
  })
  const line =
    'lcl|Scaffald_1\t80465\trs118266897\tR\tA\t29\tPASS\tNS=3;0,14;AF=0.5;DB;112;PG2.1'

  const variant = parser.parseLine(line)

  const f = new VcfFeature({
    parser,
    variant,
    id: 'myuniqueid',
  })
  expect(f.id()).toEqual('myuniqueid')
  expect(f.get('name')).toEqual('rs118266897')
})

test('try INS feature with END less than start', () => {
  const parser = new VcfParser({
    header:
      '#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tBAMs/caudaus.sorted.sam',
  })
  const line = 'chr1\t100\trs123\tR\tA\t29\tPASS\tEND=1;SVTYPE=INS'

  const variant = parser.parseLine(line)

  const f = new VcfFeature({
    parser,
    variant,
    id: 'myuniqueid',
  })
  expect(f.id()).toEqual('myuniqueid')
  expect(f.get('start')).toEqual(99)
  expect(f.get('end')).toEqual(100)
})

test('try DEL feature with END info field valid', () => {
  const parser = new VcfParser({
    header:
      '#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tBAMs/caudaus.sorted.sam',
  })
  const line = 'chr1\t100\trs123\tR\t<DEL>\t29\tPASS\tEND=1000;SVTYPE=DEL'

  const variant = parser.parseLine(line)

  const f = new VcfFeature({
    parser,
    variant,
    id: 'myuniqueid',
  })
  expect(f.id()).toEqual('myuniqueid')
  expect(f.get('start')).toEqual(99)
  expect(f.get('end')).toEqual(1000)
})

test('multiple SVs', () => {
  const parser = new VcfParser({
    header:
      '#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tBAMs/caudaus.sorted.sam',
  })
  const line =
    'chr1\t100\trs123\tR\t<INVDUP>,<INV>\t29\tPASS\tEND=1000;SVTYPE=DEL'

  const variant = parser.parseLine(line)

  const f = new VcfFeature({
    parser,
    variant,
    id: 'myuniqueid',
  })
  expect(f.get('description')).toEqual('<INVDUP>,<INV>')
})
test('BND', () => {
  const parser = new VcfParser({
    header:
      '#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tBAMs/caudaus.sorted.sam',
  })
  const line =
    'chr1\t100\trs123\tR\tG[ctgA:34200[\t29\tPASS\tEND=1000;SVTYPE=BND'

  const variant = parser.parseLine(line)

  const f = new VcfFeature({
    parser,
    variant,
    id: 'myuniqueid',
  })
  expect(f.get('description')).toEqual('G[ctgA:34200[')
})
test('multiple BND', () => {
  const parser = new VcfParser({
    header:
      '#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tBAMs/caudaus.sorted.sam',
  })
  const line =
    'chr1\t100\trs123\tR\tG[ctgA:34200[,G[ctgA:44200[\t29\tPASS\tEND=1000;SVTYPE=BND'

  const variant = parser.parseLine(line)

  const f = new VcfFeature({
    parser,
    variant,
    id: 'myuniqueid',
  })
  expect(f.get('description')).toEqual('G[ctgA:34200[,G[ctgA:44200[')
})
test('multiple SNV', () => {
  const parser = new VcfParser({
    header:
      '#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tBAMs/caudaus.sorted.sam',
  })
  const line = 'chr1\t100\trs123\tG\tA,C\t29\tPASS\tHELLO=world'

  const variant = parser.parseLine(line)

  const f = new VcfFeature({
    parser,
    variant,
    id: 'myuniqueid',
  })
  expect(f.get('description')).toEqual('SNV G -> A,C')
})

test('multiple SNV2', () => {
  const parser = new VcfParser({
    header:
      '#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tBAMs/caudaus.sorted.sam',
  })
  const line = 'chr1\t100\trs123\tG\tAT,<*>\t29\tPASS\tHELLO=world'

  const variant = parser.parseLine(line)

  const f = new VcfFeature({
    parser,
    variant,
    id: 'myuniqueid',
  })
  expect(f.get('description')).toEqual('insertion G -> AT,<*>')
})

// see example 1.1 in VCF 4.3 spec, indicates the . in ALT field indicates
// "a site that is called monomorphic reference (i.e. with no alternate alleles"
test('null ALT', () => {
  const parser = new VcfParser({
    header:
      '#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tBAMs/caudaus.sorted.sam',
  })
  const line = 'chr1\t100\trs123\tG\t.\t29\tPASS\tHELLO=world'

  const variant = parser.parseLine(line)

  const f = new VcfFeature({
    parser,
    variant,
    id: 'myuniqueid',
  })
  expect(f.toJSON()).toMatchSnapshot()
})
