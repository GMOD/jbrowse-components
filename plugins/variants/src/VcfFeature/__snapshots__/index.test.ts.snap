// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`null ALT 1`] = `
{
  "ALT": null,
  "CHROM": "chr1",
  "FILTER": "PASS",
  "ID": [
    "rs123",
  ],
  "INFO": {
    "HELLO": [
      "world",
    ],
  },
  "POS": 100,
  "QUAL": 29,
  "REF": "G",
  "aliases": undefined,
  "description": "no alternative alleles",
  "end": 100,
  "name": "rs123",
  "refName": "chr1",
  "samples": {},
  "start": 99,
  "type": "remark",
  "uniqueId": "myuniqueid",
}
`;
